### PR TITLE
Begin transformation of Eigen structure

### PR DESCRIPTION
**Pull request by btracey**

### CODE REVIEW

**Comment by btracey** [comment]
Am I reading the code properly on e?

File: mat64/eigen.go [hunk 1]
```diff
@@ -23,12 +23,61 @@ func symmetric(m *Dense) bool {
...
+	r, c := a.Dims()
+	if r != c {
+		panic(matrix.ErrShape)
+	}
+	aCopy := DenseCopyOf(a)
+	e.vectorsComputed = vectors
+
+	e.ef = eigen(aCopy, 0)
+	e.n = r
+	return ok
```
Comment (kortschak) [inline mat64/eigen.go]: `return true`?

Comment (btracey) [inline mat64/eigen.go]: Whoops. Done.

**Comment by kortschak** [comment]
>Am I reading the code properly on e?

What do you mean?

**Comment by btracey** [comment]
Are the values of e actually in positive negative pairs? Do these values need sorting like in D(), or is as is fine?

**Comment by kortschak** [comment]
Not sure.

**Comment by kortschak** [comment]
Making a quick test on the test data (change the sign -2 on L26 in the tests), what you are doing is right.

File: mat64/eigen.go [hunk 2]
```diff
@@ -23,12 +23,61 @@ func symmetric(m *Dense) bool {
 	return true
 }
 
-type EigenFactors struct {
+// Eigen is a type for creating an using the eigenvalue decomposition of a amtrix.
```
Comment (vladimir-ch) [inline mat64/eigen.go]: s/an using/and using/

File: mat64/eigen.go [hunk 3]
```diff
@@ -23,12 +23,61 @@ func symmetric(m *Dense) bool {
...
+
+	ef eigenFactors
+}
+
+// Factorize computes the eigenvalue decompositon of the input square matrix a.
+// The eigenvalue decomposition is defined as
+//  A = P * D * P^-1
+// where D is a diagonal matrix containing the eigenvalues of the matrix, and
+// P is a matrix of the eigenvectors of A. If the vectors input argument is
+// false, the eigenvectors are not computed.
```
Comment (vladimir-ch) [inline mat64/eigen.go]: This is not true at the moment. `vectors` is assigned to `vectorsComputed` which is then never used.

Comment (btracey) [inline mat64/eigen.go]: Yes. The whole function needs to be replaced with a lapack implementation.

Comment (btracey) [inline mat64/eigen.go]: I can add a BUG comment or something.

File: mat64/eigen.go (hunk 2 above)
Comment (kortschak) [inline mat64/eigen.go]: s/amtrix/matrix/

File: mat64/eigen.go [hunk 4]
```diff
@@ -23,12 +23,61 @@ func symmetric(m *Dense) bool {
 	return true
 }
 
-type EigenFactors struct {
+// Eigen is a type for creating an using the eigenvalue decomposition of a amtrix.
+type Eigen struct {
+	vectorsComputed bool
+
+	n int
```
Comment (kortschak) [inline mat64/eigen.go]: Comment on this field.

File: mat64/eigen.go [hunk 5]
```diff
@@ -23,12 +23,61 @@ func symmetric(m *Dense) bool {
...
+
+	ef eigenFactors
+}
+
+// Factorize computes the eigenvalue decompositon of the input square matrix a.
```
Comment (kortschak) [inline mat64/eigen.go]: s/eigenvalue decomposition/Eigen decomposition/

File: mat64/eigen.go [hunk 6]
```diff
@@ -23,12 +23,61 @@ func symmetric(m *Dense) bool {
...
+// The eigenvalue decomposition is defined as
```
Comment (kortschak) [inline mat64/eigen.go]: Same here.

File: mat64/eigen.go [hunk 7]
```diff
@@ -23,12 +23,61 @@ func symmetric(m *Dense) bool {
...
+	aCopy := DenseCopyOf(a)
+	e.vectorsComputed = vectors
+
+	e.ef = eigen(aCopy, 0)
+	e.n = r
+	return ok
+}
+
+// Vectors extracts the eigenvalues of the factorized matrix. If the input is
+// non-nil, the values are stored in-place into the receiver. In this case the
```
Comment (kortschak) [inline mat64/eigen.go]: 'receiver'?

File: mat64/eigen.go [hunk 8]
```diff
@@ -23,12 +23,61 @@ func symmetric(m *Dense) bool {
...
+	return v
+}
+
+type eigenFactors struct {
 	V    *Dense
 	d, e []float64
 }
 
-// Eigen returns the Eigenvalues and eigenvectors of a square real matrix.
+// eigen returns the Eigenvalues and eigenvectors of a square real matrix.
```
Comment (kortschak) [inline mat64/eigen.go]: lc/Eigenvalues/

**Comment by kortschak** [comment]
LGTM after comments addressed.

File: mat64/eigen.go (hunk 2 above)
Comment (btracey) [inline mat64/eigen.go]: Done.

File: mat64/eigen.go (hunk 4 above)
Comment (btracey) [inline mat64/eigen.go]: Done.

File: mat64/eigen.go (hunk 5 above)
Comment (btracey) [inline mat64/eigen.go]: Done.

File: mat64/eigen.go (hunk 6 above)
Comment (btracey) [inline mat64/eigen.go]: Done.

File: mat64/eigen.go (hunk 3 above)
Comment (btracey) [inline mat64/eigen.go]: Done.

File: mat64/eigen.go (hunk 7 above)
Comment (btracey) [inline mat64/eigen.go]: Done.

File: mat64/eigen.go (hunk 8 above)
Comment (btracey) [inline mat64/eigen.go]: Done.

**Pull request merged**